src/Control/Monad/Rec/Class.purs

```diff
@@ -1,27 +1,33 @@
 module Control.Monad.Rec.Class
-  ( class MonadRec
+  ( Step(..)
+  , class MonadRec
   , tailRec
   , tailRecM
   , tailRecM2
   , tailRecM3
   , forever
   ) where
 
-import Prelude (class Monad, unit, (<$), (<$>), ($), pure, bind, (<<<))
+import Prelude
 
-import Control.Monad.Eff (Eff(), untilE)
+import Control.Monad.Eff (Eff, untilE)
 import Control.Monad.Eff.Unsafe as U
-import Control.Monad.ST (ST(), runST, newSTRef, readSTRef, writeSTRef)
+import Control.Monad.ST (ST, runST, newSTRef, readSTRef, writeSTRef)
 
-import Data.Either (Either(..), fromRight)
+import Data.Either (Either(..))
 import Data.Identity (Identity(..))
 
 import Partial.Unsafe (unsafePartial)
 
--- | This type class captures those monads which support tail recursion in constant stack space.
+-- | The result of a computation: either `Loop` containing the updated
+-- | accumulator, or `Done` containing the final result of the computation.
+data Step a b = Loop a | Done b
```

Comment:
We might want a `Functor`/`Bifunctor` instance here, I have a feeling it will be useful when implementing the `transformers` instances.